pkg/cache/informer_cache.go

```diff
@@ -152,6 +153,24 @@ func (ic *informerCache) GetInformer(ctx context.Context, obj client.Object) (In
 	return i.Informer, err
 }
 
+// GetInformerNonBlocking returns the informer for the obj without waiting for its cache to sync.
+func (ic *informerCache) GetInformerNonBlocking(ctx context.Context, obj client.Object) (Informer, error) {
```

Comment:
I might be missing something, where is the new code going to be used and how?

Reply:
Got it, thanks for the example. I think overall the approach seems good, there might be more feedback in terms of API changes and whatnot.

I assume to actually use these methods we'll require changes to the exposed interface?

For the non-blocking scenario, could we opt into an option instead of adding a new method?

Reply:
For the controller runtime project specifically, it might be worth discussing how we'll use these facilities to shutdown informers at runtime

cc @alvaroaleman

Reply:
Totally cool with adding an option. Are you thinking of:

* adding functional options (e.g. `GetInformer(ctx, obj, Blocking(false), SomeOtherOption())`)
* Adding a straight boolean: `GetInformer(ctx, obj, false)`
* An options object: `GetInformer(ctx, obj, Options{Blocking: false})`

?

Reply:
Probably the first option would be more in line with what we have elsewhere, I was thinking something pretty simple like:

```golang
// Informers knows how to create or fetch informers for different
// group-version-kinds, and add indices to those informers.  It's safe to call
// GetInformer from multiple threads.
type Informers interface {
	// GetInformer fetches or constructs an informer for the given object that corresponds to a single
	// API kind and resource.
	GetInformer(ctx context.Context, obj client.Object, opts ...InformerOption) (Informer, error)

	// GetInformerForKind is similar to GetInformer, except that it takes a group-version-kind, instead
	// of the underlying object.
	GetInformerForKind(ctx context.Context, gvk schema.GroupVersionKind, opts ...InformerOption) (Informer, error)
[...]
}
```

And ultimately define the option such as `NonBlocking()` or similar

Reply:
The above is a breaking change, but it provides callers with a backward compatible approach, and allows us to extend the option further in the future if needed

Reply:
Functional options it is!

Reply:
@vincepri @maxsmythe I don't think the "Blocking" option will be compatible with reference counting the number of calls to AddEventHandler and RemoveEventHandler. If no calls to AddEventHandler have been made yet, we won't start the informer yet. Only once we have one or more `ResourceEventHandlerRegistration`s, we start the informer. So, I would instead opt for removing the `Blocking` option.
I think DynamicSharedInformerFactory took the same approach (since no ctx is passed to ForResource (see https://github.com/kubernetes/kubernetes/blob/589d6f388680b29f7f7961bda4f4262733b54210/staging/src/k8s.io/client-go/dynamic/dynamicinformer/interface.go#L32).

Reply:
@inteon blocking is actually the current behavior. This option allows us to get an informer without blocking-for-sync. As such, I think adding this option is moving things in the direction you're suggesting.

blocking is still necessary for supporting `client.Get()`, which is synchronous and assumes the informer cache is ready to be read from.

Does that allay your concern?

Reply:
closing this, since non-blocking get has been moved to #2371